cmd/k8s-operator/logger.go

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !plan9
+
+package main
+
+import (
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	kzap "sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"tailscale.com/types/logger"
+)
+
+type (
+	logfSink struct {
+		logf logger.Logf
+	}
+)
+
+// wrapZapCore returns a zapcore.Core implementation that splits the core chain using zapcore.NewTee. This causes
+// logs to be simultaneously written to both the original core and the provided logger.Logf function.
+func wrapZapCore(core zapcore.Core, logf logger.Logf) zapcore.Core {
+	// We use a tee logger here so that logs are written to stdout/stderr normally while at the same time being
+	// sent upstream.
+	return zapcore.NewTee(core, zapcore.NewCore(&kzap.KubeAwareEncoder{
+		Encoder: zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()),
+		Verbose: true,
+	}, &logfSink{logf: logf}, zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
+		return true
+	})))
+}
+
+func (l *logfSink) Write(p []byte) (n int, err error) {
+	l.logf("k8s: %s", p)
```

Comment:
This seems like the wrong place to insert this context to me. If we want to make it clear that some logs are from the reconcilers, we can put that in the logger fields. But I think it's already pretty clear which logs are from the reconcilers anyway because we put the reconciler name in each logger, so I'd prefer to just write the buffer we're passed without modification.

Reply:
The main motivation for this here is so that within logz one can use a search term to filter exclusively for logs from the operator. I suppose at the top level of the logger we can add a field that will propagate to all reconciler logs? If we think that's necessary?

I found that because the full breadth of logs can be quite noisy, it was nice to be able to just show operator logs.